### DISCUSSION TITLE
Random values based on space and time

### DISCUSSION THREAD

**Discussion by batodon** [idaholab/moose#22971] [Q&A General] [answered]
Hello,

Is there a function in moose to generate a random value as a function of space and time (i.e., `[_qp]` and `_dt`)? In other words, the function will take the arguments of space and time. I tried using the c++ `rand()` function, but it only gives a random value between two extreme values. Is there any moose example having such a function? Any help is deeply appreciated. Thanks.

**Comment by GiudGiud** [comment, marked as answer]
You ll want to have a look at the random interface
https://mooseframework.inl.gov/source/interfaces/RandomInterface.html
https://mooseframework.inl.gov/docs/doxygen/moose/classRandomInterface.html

**Reply by GiudGiud** [reply]
the randomIC is an example using it

**Reply by batodon** [reply]
Hi @GiudGiud. Thanks for the suggestion. I took a look at the `RandomInterface` class. It seems the `getRandomReal()` or `getRandomLong() `functions generate the same random numbers for ‘all’ elements within the domain at a ‘snapshot.’ I mean, all the elements within the domain will have the same random numbers at a snapshot. What I’m looking for, though, is a function that generates different random numbers for all the different elements at a snapshot (or single step). That is, element 1 should have a random value different from element 2 at a single step. Thanks.

**Reply by GiudGiud** [reply]
it should not be like that. have a look at RandomIC, it s spatially random

**Reply by batodon** [reply]
Gotcha! Thanks, a lot @GiudGiud.

**Reply by batodon** [reply]
Hello @GiudGiud 
I had to re-open this ticket since I did not achieve success with my simulation. Maybe, I did not properly communicate my objective for using the random values. What I really want to achieve is to rotate each element using a unique random angle of rotation for that element. I had previously rotated my model using a single rotation angle for all elements. However, I find it challenging to do the same thing using different rotation angles for all elements simultaneously. Yes, the `RandomIC` generates spatially varied field data for a given variable, but I wonder how I can apply this to my objective. The following code shows how I’m using the `RandomInterface` class, but this approach only generates a single random rotation angle at a time. I believe this is incorrect, considering my objective. Please, do you have any suggestions on how to approach this problem? Thanks. 

```
// Whether to use spatially random angle of rotation at each timestep or a fixed angle of rotation.
    Real  _randm_rad_xy;
    Real  _randm_rad_yz;
      if (_distribution)
        {
          _randm_rad_xy = _distribution->quantile(getRandomReal());
          _randm_rad_yz = _distribution->quantile(getRandomReal());
        }
      else
          {
            _randm_rad_xy = getRandomReal() * (_max - _min) + _min;
            _randm_rad_yz = getRandomReal() * (_max - _min) + _min;
          }

      Real  _rad_xy;
      Real  _rad_yz;
          if (_Random_field)
        // Get the spatially random rotation angle (in radians) for each element at each timestep.
          {
            _rad_xy = _randm_rad_xy;
            _rad_yz = _randm_rad_yz;
          }
          else
        // Get the fixed (or user-specified) rotation angle for all elements in the domain at each timestep.
          {
          _rad_xy = _fix_rad_xy;
          _rad_yz = _fix_rad_yz;
          }
```

**Reply by GiudGiud** [reply]
Hello

I did not mean for you to use the RandomIC, sorry for the misunderstanding.
I meant for you to look at the code in the RandomIC to extract the appropriate API calls from the randomInterface()

It uses
```
    return generateRandom() * (_max - _min) + _min;
```
This `generateRandom()` is different at every quadrature point. I think it should be different every time you call it, so it can be different in time, in space etc. 

Does this work for you?

**Reply by batodon** [reply]
Hi @GiudGiud. Yeah, I got a bit confused. I'll work around it and let you know shortly. Thanks.

**Reply by batodon** [reply]
I fixed it! It compiles, beautifully...
